feat: dropdown component

deboer-tim opened the pull request:
### What does this PR do?

As per the last time this item was discussed, I spent a while trying to use svelte-select and a few other off-the-shelf select components to work, but ultimately they all had the same problem as we had with pulling images: even after tying to style them, I couldn't get any of them to fit into our current styling well enough that I was comfortable submitting a PR.

Two other minor things to note:
- Most components used some form of item[] instead of child option elements. This might be useful in the future (and could be added here) but would have required more migration up front.
- We don't have any immediate plans to expose this component in the UI library, but using a third-party library would have excluded that possibility.

So, here it is instead: I used the code from the Typeahead component, mixing in bits from Tooltip and other places to create a custom styled Dropdown component. I tried to keep required migration to a minimum, but some tests needed to be updated and I switched onInput to svelte 5 style.

Updated the preferences page and Run Image form to use the component so that there is something to test. There are ~13 other places where we use select which would be migrated in follow-up PRs.

### Screenshot / video of UI

https://github.com/user-attachments/assets/a908e81b-4ac9-406f-b0bc-ac5b75b32702

### What issues does this PR fix or reference?

Core of issue #6199.

### How to test this PR?

Go to Settings > Preferences and Run Image form.

- [x] Tests are covering the bug fix or the new feature

axel7083 commented:
> We don't have any immediate plans to expose this component in the UI library

Actually I would have though we should do it the other way, first add it the ui library and create a Storybook page for it, so we can experiment with it.

> but using a third-party library would have excluded that possibility.

We are exposing the following dependencies through the @podman-desktop/ui-svelte library

https://github.com/containers/podman-desktop/blob/8e268d8df7f75422ace9cd396d4c4045de1344f0/packages/ui/package.json#L129-L135

---

![image](https://github.com/user-attachments/assets/0d41573c-08a2-4526-a5de-0700d5e2b816)

I don't think the popup should be over the form itself.

If we look at some libraries we can clearly see they put is bellow.

![Screenshot from 2024-10-02 09-46-11](https://github.com/user-attachments/assets/838f1b29-b334-47a2-a087-b816268b8df0)

gastoner commented on packages/renderer/src/lib/ui/Dropdown.svelte:
```diff
@@ -0,0 +1,201 @@
+    <span class="grow">{selectLabel}</span>
+    <div
+      class:text-[var(--pd-input-field-stroke)]={!disabled}
+      class:text-[var(--pd-input-field-disabled-text)]={!!disabled}
```
Shouldn't be here `!` instead of `!!` ?

deboer-tim replied:
Yes, good catch. Typo fixed.

gastoner commented:
Also, it is probably a different issue; on the Proxy page, we are not using the same dropdown (which we probably should?). 
![image](https://github.com/user-attachments/assets/e354e622-152b-426e-8b95-8a82f560011f)

And +1 to Axels comment with popup over form

deboer-tim commented:
> > We don't have any immediate plans to expose this component in the UI library
> 
> Actually I would have though we should do it the other way, first add it the ui library and create a Storybook page for it, so we can experiment with it.
> 
> > but using a third-party library would have excluded that possibility.
> 
> We are exposing the following dependencies through the @podman-desktop/ui-svelte library

I thought we took a decision that we would keep that list short, i.e. we provide basic controls to match PD's style, but if an extension wants something more complex like a chart we would not provide it; extension should use and style third-party library directly.

I get what you're saying about directly exposing in the UI library, but I would prefer to be more cautious, i.e. we add a component, make sure it works for all our use cases, go through at least one QE cycle, confirm there's an extension that wants/will use it, then expose. I would assume it's safe enough in this case (?) but didn't want to presume or overload a PR.

> I don't think the popup should be over the form itself.
> 
> If we look at some libraries we can clearly see they put is bellow.

I was trying to match the native one, which is apparently the only one that goes over top. I've moved the popup down, and since you can now click the control again that will now close it.

deboer-tim commented:
> Also, it is probably a different issue; on the Proxy page, we are not using the same dropdown (which we probably should?).

Correct, there are >15 HTML select components and only 4 (preferences enums + 3 in Run Image form) are fixed here to have something to test. Another PR or two will be required to switch all the others and mark #6199 complete.

> And +1 to Axels comment with popup over form

👍🏼 PR updated.

jeffmaury reviewed (approved):
LGTM but I don't get why teh component did not land in UI library

axel7083 commented on packages/renderer/src/lib/ui/Dropdown.svelte:
```diff
@@ -0,0 +1,209 @@
+
+  {#if opened}
+    <div
+      class="absolute top-full right-0 z-10 w-full max-h-80 rounded bg-[var(--pd-dropdown-bg)] border-[var(--pd-input-field-hover-stroke)] border-[1px] overflow-y-auto whitespace-nowrap">
```
Here we are using `rounder`, and above we are using `rounded-md`

https://github.com/containers/podman-desktop/blob/70d2e3604d71ab5b867f682e62af733856afafc7/packages/renderer/src/lib/ui/Dropdown.svelte#L156

![image](https://github.com/user-attachments/assets/07c68fde-b544-4451-984f-c568c52472a9)

And some component above are using `rounder-sm`

https://github.com/containers/podman-desktop/blob/70d2e3604d71ab5b867f682e62af733856afafc7/packages/renderer/src/lib/image/RunImage.svelte#L853-L854

Should we make them match ?

deboer-tim replied:
Updated both to rounded-md, which is what design uses.

benoitf reviewed (dismissed):
IMHO the PR should only focus on bringing the component and not bringing the component AND changing the usage of other parts using it.

At some point, it was hard to demonstrate the component without experiencing it. But now, we have a way to showcase a component, this is why we made storybook stories. You can do trial and live change to the component and be focused only on the component.

Also to me the end goal is to have in the UI library so if it's first in renderer and we change the code to use it there, then we move it to UI, it would make additional PR to replace internal to external that we can prevent doing it only once.

We could flag the component as being in 'beta' state

deboer-tim commented:
Several people commented on going straight to UI component, so I moved it there. Added some typing due to higher checks in ui package, added a basic story, and removed the changes in renderer. Rebased to clear up multiple changes to renderer + removal, and get on latest.

Fixed inconsistent rounded-md.

Not sure how to mark as 'beta' or whether it's worth it.

axel7083 commented:
> Not sure how to mark as 'beta' or whether it's worth it.

Dropdown can be tested here https://66fea7a62443dcad09ce6f63--podman-desktop-pr.netlify.app/storybook?id=dropdown--docs

benoitf commented on packages/ui/src/lib/dropdown/Dropdown.svelte:
```diff
@@ -0,0 +1,210 @@
+import { faCaretDown, faCheck } from '@fortawesome/free-solid-svg-icons';
+import Fa from 'svelte-fa';
+
+export let id: string | undefined = undefined;
+export let name: string | undefined = undefined;
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export let value: any = '';
+export let disabled: boolean = false;
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export let onInput: (val: any) => void = () => {};
+
+interface Option {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
I don't know why we have so many 'any' there
it looks like we're having more than 10 // eslint-disable-next-line @typescript-eslint/no-explicit-any


should we have a generics for the type of the field we want to handle ?
or is it just `unknown`

deboer-tim replied:
Most should be unknown, updated.

benoitf commented on packages/ui/src/lib/dropdown/Dropdown.svelte:
```diff
@@ -0,0 +1,210 @@
+export let value: any = '';
+export let disabled: boolean = false;
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export let onInput: (val: any) => void = () => {};
```
it's not clear to me what the `onInput` will do, is it more accurate to use `onChange` ?

deboer-tim replied:
I had used onInput since that's what select had, but switched to more consistent onChange instead.

benoitf commented on packages/ui/src/lib/dropdown/Dropdown.svelte:
```diff
@@ -0,0 +1,210 @@
+import { faCaretDown, faCheck } from '@fortawesome/free-solid-svg-icons';
+import Fa from 'svelte-fa';
+
+export let id: string | undefined = undefined;
+export let name: string | undefined = undefined;
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export let value: any = '';
```
IMHO you should use $props for new components as reactivity is better handled (it can say for example if we're missing some $state on some fields, etc)

deboer-tim replied:
Updates to props. As per comment, this triggered requiring everything to Svelte 5.

benoitf commented on packages/ui/src/lib/dropdown/Dropdown.svelte:
```diff
@@ -0,0 +1,210 @@
+let pageStep: number = 10;
+let comp: HTMLElement;
+
+$: selectLabel = options.find(o => o.value === value)?.label ?? value;
```
==> $effect

benoitf replied:
and options should be a state then

deboer-tim replied:
Fixed.

benoitf commented on storybook/src/stories/Dropdown.stories.svelte:
```diff
@@ -0,0 +1,62 @@
+      defaultValue: undefined,
+    },
+  },
+});
```
I think it's not telling the callback like
https://github.com/containers/podman-desktop/blob/main/storybook/src/stories/button/Button.stories.svelte#L17-L19

deboer-tim replied:
Added onChange function to Storybook.

benoitf commented on storybook/src/stories/Dropdown.stories.svelte:
```diff
@@ -0,0 +1,62 @@
+    },
+    name: {
+      control: 'text',
+      description: 'Name of the dropdown',
```
where the name is used, is it for just an aria-label or something else

deboer-tim replied:
Just the control name, used by some playwright tests. Looks like we aren't adding it to Storybook in most other cases so I removed it.

deboer-tim commented:
Rebased on main and made changes based on review comments. Adding some Svelte 5 kept triggering more changes required so it is now fully updated to Svelte 5.
- any fixed to unknown and eslint disables removed
- use Svelte 5 props, bindable, effect, state
- use Svelte 5 render/snippets
- use Svelte 5 events / functions
- updated Storybook. Removed id/name (inconsistent, but we seem to omit simple/aria props for most components) and added options.

Ready for another review.
